torchtune/modules/common_utils.py

```diff
@@ -442,3 +448,131 @@ def delete_kv_caches(model: nn.Module):
         if hasattr(module, "kv_cache") and callable(module.kv_cache):
             module.cache_enabled = False
             module.kv_cache = None
+
+
+def resize_token_embeddings(model: nn.Module, num_embeddings: int) -> None:
+    """
+    Resizes the token embeddings and the final output projection layer of a ``TransformerDecoder`` model.
+    The default init strategy is taking the mean of all the embeddings, new embeddings will be
+    instantiated to this value.
+
+    This function modifies the model in-place.
+
+    The primary purpose is to adjust the vocabulary size of a pre-trained model.
+    This is useful when fine-tuning a model on a dataset with a different
+    vocabulary or when adding special tokens.
+
+    Example:
+        >>> model = setup_model(...)
+        >>> tokenizer = setup_tokenizer(...)
+        >>> resize_token_embedding(model, tokenizer.num_embeddings)
+
+    Args:
+        model (nn.Module): The transformer model to modify. The model is
+            expected to have ``tok_embeddings`` (an ``nn.Embedding`` layer) and
+            ``output`` (e.g., ``nn.Linear`` or ``TiedLinear``) attributes.
+        num_embeddings (int): The desired number of embeddings in the resized
+            embedding layer and output projection layer.
+
+    Returns:
+        None: The function modifies the `model` in-place.
+
+    Raises:
+        AssertionError: When model not of type ``TransformerDecoder`` is passed in
+            and is a resize is needed. User should pass in `TransformerDecoder` part
+            of the model only.
+    """
+    need_resize = _need_to_resize(model, num_embeddings)
+    if not need_resize:
+        _log.info("Embedding resize not needed, skipping")
+        return
+
+    if not isinstance(model, TransformerDecoder) and need_resize:
+        raise AssertionError(
+            f"Resize needed but model is of type {type(model)}, please pass in TransformerDecoder to function"
+        )
```

Comment:
Something like this? @pbontrager @iamzainhuda 

```suggestion
    if isinstance(model, DeepFusionModel) or isinstance(model, EarlyFusionModel):
        model = model.decoder
    if isinstance(model.tok_embeddings, FusionEmbedding) and resize_needed:
        raise AssertionError(
            f"Resize is not supported for FusionEmbeddings"
        )
```